core/modules/loader/MWorkerList.cc

```diff
@@ -0,0 +1,424 @@
+    LOGS(_log, LOG_LVL_DEBUG, "MWorkerListItem::addDoListItems");
+    std::lock_guard<std::mutex> lck(_doListItemsMtx);
+    if (_sendListToWorker == nullptr) {
+        _sendListToWorker = std::make_shared<SendListToWorker>(shared_from_this(), _central);
+        _central->addDoListItem(_sendListToWorker);
+    }
+    if (_reqWorkerKeyInfo == nullptr) {
+        _reqWorkerKeyInfo = std::make_shared<ReqWorkerKeyInfo>(shared_from_this(), _central);
+        _central->addDoListItem(_reqWorkerKeyInfo);
+    }
+}
+
+
+void MWorkerListItem::flagNeedToSendList() {
+    auto slw = _sendListToWorker;
+    if (slw != nullptr) { slw->setNeedInfo(); }
+}
+
+
+void MWorkerListItem::sendListToWorkerInfoReceived() {
+    auto slw = _sendListToWorker;
+    if (slw != nullptr) {
+        // _sendListToWorker is a tough one to tell if the worker got the info, so
+        // it is assumed that this worked when the list is sent. The worker
+        // will either ask for it or it will be sent again later.
+        slw->infoReceived();
+    }
+}
+
+
+void MWorkerListItem::setRangeStr(StringRange const& strRange) {
+    std::lock_guard<std::mutex> lck(_mtx);
+    _range = strRange;
+}
+
+
+void MWorkerListItem::setAllInclusiveRange() {
+    LOGS(_log, LOG_LVL_INFO, "MWorkerListItem::setAllInclusiveRange for name=" << _wId);
+    std::lock_guard<std::mutex> lck(_mtx);
+    _range.setAllInclusiveRange();
+    _active = true;  /// First worker.
+}
+
+
+void MWorkerListItem::setNeighborsInfo(NeighborsInfo const& nInfo) {
+    std::lock_guard<std::mutex> lck(_mtx);
+    _neighborsInfo.keyCount = nInfo.keyCount;
+    _neighborsInfo.recentAdds = nInfo.recentAdds;
+
+    auto old = _neighborsInfo.neighborLeft->get();
+    if (old != 0) {
+        LOGS(_log, LOG_LVL_WARN, "Worker=" << _wId <<
+                "neighborLeft changing from valid old=" << old <<
+                " to new=" << nInfo.neighborLeft->get());
+    }
+    if (old != nInfo.neighborLeft->get()) {
+        LOGS(_log, LOG_LVL_INFO, "Worker=" << _wId <<
+                "neighborLeft=" << nInfo.neighborLeft->get());
+    }
+    _neighborsInfo.neighborLeft->update(nInfo.neighborLeft->get());
+
+
+    old = _neighborsInfo.neighborRight->get();
+    if (old != 0) {
+        LOGS(_log, LOG_LVL_WARN, "Worker=" << _wId <<
+                "neighborRight changing from valid old=" << old <<
+                " to new=" << nInfo.neighborRight->get());
+    }
+    if (old != nInfo.neighborRight->get()) {
+
+        LOGS(_log, LOG_LVL_INFO, "Worker=" << _wId <<
+                "neighborRight=" << nInfo.neighborRight->get());
+    }
+    _neighborsInfo.neighborRight->update(nInfo.neighborRight->get());
+}
+
+
+int MWorkerListItem::getKeyCount() const {
+    return _neighborsInfo.keyCount;
+}
+
+
+std::ostream& operator<<(std::ostream& os, MWorkerListItem const& item) {
+    os << "name=" << item._wId << " address=" << *item._udpAddress << " range(" << item._range << ")";
```

Comment:
Is this thread safe? My understanding is that worker's ranges may change, and the range is ot a trivial data type.

Reply:
other places that use _range take a lock. That should be done here too I think?